qark/plugins/cert/cert_validation_methods_overriden.py

```diff
@@ -0,0 +1,94 @@
+                    "You may need to toggle the proxy on/off to get past sections that do validate certificates "
+                    "properly in order to reach the vulnerable code. This proves that it will accept certificates "
+                    "from any CA. You should always validate your configuration by visiting an HTTPS site in the "
+                    "native browser and verifying you receive a certificate warning. For details, please see: "
+                    "https://developer.android.com/training/articles/security-ssl.html")
+
+
+class CertValidation(BasePlugin):
+    """
+    This plugin checks if a method in `CERT_METHODS` is overriden with an insecure version that usually does not verify
+    SSL connections.
+    """
+    def __init__(self):
+        BasePlugin.__init__(self, category="cert")
+        self.severity = Severity.WARNING
+
+    def _process_file(self, filepath):
+        try:
+            with open(filepath, 'r') as f:
+                file_contents = f.read()
+        except IOError:
+            log.debug("Unable to read file")
+            return
+
+        try:
+            tree = javalang.parse.parse(file_contents)
+        except javalang.parser.JavaSyntaxError:
+            log.debug("Couldn't parse the java file: %s", filepath)
+            return
+
+        cert_methods = (method_declaration for _, method_declaration in tree.filter(MethodDeclaration)
+                        if method_declaration.name in CERT_METHODS)
+        for cert_method in cert_methods:
+            if cert_method.name == "checkServerTrusted":
+                self._check_server_trusted(cert_method)
+            elif cert_method.name == "onReceivedSslError":
+                self._on_received_ssl_error(cert_method)
+
+    def _check_server_trusted(self, cert_method):
+        """
+        Determines if the `checkServerTrusted` method is overriden with either no function body (defaults to allow)
+        or if the body just returns.
+        """
+        if not cert_method.body:
+            self.issues.append(Issue(category=self.category, name="Empty certificate method",
+                                     severity=self.severity,
+                                     description=CHECK_SERVER_TRUSTED + MITM_DESCRIPTION))
+        elif len(cert_method.body) == 1 and type(cert_method.body[0]) is ReturnStatement:
+            self.issues.append(Issue(category=self.category, name="Empty (return) certificate method",
+                                     severity=self.severity,
+                                     description=CHECK_SERVER_TRUSTED + MITM_DESCRIPTION))
+
+    def _on_received_ssl_error(self, cert_method):
+        """
+        Determines if the `onReceivedSslError` method is overriden with a body of `handler.proceed()` which will
+        drop SSL errors.
+        """
+        if len(cert_method.body) == 1 and type(cert_method.body[0]) is StatementExpression:
```

Comment:
Thinking this should recursively check nodes, in case someone is logging and then overriding

Reply:
This is a complex issue, because there are a lot of ways around a check like that. There are a lot of nodes in the AST that can be considered no-ops but I don't know an easy way to check that a statement is going to be hit 100% of the time besides checks similar to this one. I'm open to a way where we can expand detection of this method but I can't recognize a clear path.